transport/ssh.go

```diff
@@ -47,7 +47,7 @@ func getHostKeyCallback() (ssh.HostKeyCallback, error) {
 	return knownhosts.New(filepath.Join(homedir, ".ssh", "known_hosts"))
 }
 
-func sshConnect(host, port, user string, strictHostKeyChecking bool) (*ssh.Client, error) {
+func sshConnect(host, port, user string, identityfile string, strictHostKeyChecking bool) (*ssh.Client, error) {
```

Comment:
identityfile appears to be unused. Was there trivial work to use it, or is that something you're still exploring?

Reply:
Yikes, yep, thank you. Your other comment is totally related to this also, and part of the same issue really. I figured out the fix last night I think, so will be back with that change soon. Thanks for the _really_ fast review/feedback. What I should do with this (I think) is add some conditional logic for when ```sshConfig``` is getting defined. The Auth field should now be ```[]ssh.AuthMethod{ssh.PublicKeys(signer),}``` if a key name is found and ```[]ssh.AuthMethod{agent}``` if it isn't. 

For other contributors who might be reading this, my mess-up was I thought ```exec``` was connecting successfully because my code was making it see ```~/.ssh/<keyname>.pem``` But it was actually just getting the key from my running ssh-agent. IOW I was being tricked by a flawed testing cycle. You don't want to have ```AddKeysToAgent``` set to ```yes``` globally  in your default ssh config file when testing  ssh code!

This also uncovered another bug related to this. When assembling a path string, the ```+``` operator is not a good idea -- use ```filepath.Join()``` instead, so I'll fix that too.

Reply:
The Auth field takes an array, and I believe it will try all the options. If you find a valid public key, we should still pass the agent as a fallback method, i.e. `[]ssh.AuthMethod{ssh.PublicKeys(signer), agent}`.

Reply:
Makes sense, will do. Thanks!

Reply:
Here's how I approached that in my fork (not committed yet though). Which I think achieves what you said(?) though surely with a more rookie coding approach/style.
```//NEW
		authmethod := []ssh.AuthMethod{agent}
		key, err := ioutil.ReadFile(identityfile)
		if err != nil {
			log.Fatalf("unable to read private key: %v", err)
		} else {
			signer, err := ssh.ParsePrivateKey(key)
			if err != nil {
		    fmt.Errorf("unable to parse private key: %v", err)
			}
			authmethod = []ssh.AuthMethod{ssh.PublicKeys(signer),}
		}

		sshConfig := &ssh.ClientConfig{
			User:            user,
			Auth:            authmethod,
			HostKeyCallback: hostKeyCallback,
		}
//NEW
```

Reply:
I see a but in the above. Editing...

Reply:
Going to make some fixes
```
	var authmethod []ssh.AuthMethod
	if key, err := ioutil.ReadFile(identityfile); err != nil {
		activity.Record("Unable to read private key, falling back to SSH agent: %v", err)
	} else {
		if signer, err := ssh.ParsePrivateKey(key); err != nil {
			activity.Record("uUnable to parse private key, falling back to SSH agent: %v", err)
		} else {
			authmethod = append(authmethod, ssh.PublicKeys(signer))
		}
	}
	// Append agent now so that it comes last in case we find another method to try.
	authmethod = append(authmethod, agent)

	sshConfig := &ssh.ClientConfig{
		User:            user,
		Auth:            authmethod,
		HostKeyCallback: hostKeyCallback,
	}
```

Reply:
Another note: you always need to explicitly return. Calling `fmt.Errorf` just creates an error object, it doesn't do anything with it. So those lines are effectively no-ops.

Reply:
Cool, thank you!

Reply:
I implemented the changes and tested. Ready when you are for another review. 
Note that I punted on using the activity.Record function. It wasn't working here, and it turned out to be harder for me to figure out than I expected. Happy to change that, but hoping it's something simple I'm just missing. Could you explain? No rush of course.

Reply:
https://godoc.org/github.com/puppetlabs/wash/activity describes `activity.Record` a bit. I think what's missing is that Wash initializes a Journal ID on the contexts it passes to plugin methods.

Reply:
Sounds good.